ingeniamotion/communication.py

```diff
@@ -1069,3 +1094,163 @@ def boot_mode_moco(self, servo: str = DEFAULT_SERVO) -> None:
             )
         except ILError as e:
             self.logger.debug(e)
+
+    def __load_ensemble_fw_canopen(
+        self,
+        net: CanopenNetwork,
+        fw_file: str,
+        slave: int,
+        status_callback: Optional[Callable[[str], None]] = None,
+        progress_callback: Optional[Callable[[int], None]] = None,
+        error_enabled_callback: Optional[Callable[[bool], None]] = None,
+    ) -> None:
+        """Load FW to an ensemble of servos through Canopen (in parallel).
+
+        The FW files that are contained in the fw_file will be loaded by selecting any slave which
+        is part of the ensemble.
+
+        Args:
+            net: Canopen network.
+            fw_file: Path to the ensemble FW file.
+            slave: Slave ID (any slave in the ensemble)
+
+        Raises:
+            IMException: If the load FW process of any slave failed.
+        """
+        mapping = self.__unzip_ensemble_fw_file(fw_file)
+        scanned_slaves = net.scan_slaves_info()
+        first_slave_in_ensemble = self.__check_ensemble(scanned_slaves, slave, mapping)
+        threads = {}
+        with ThreadPoolExecutor() as executor:
+            for slave_id_offset, fw_file_prod_code in mapping.items():
+                slave_id = first_slave_in_ensemble + slave_id_offset
+                threads[slave_id] = executor.submit(
+                    net.load_firmware,
+                    slave_id,
+                    fw_file_prod_code[0],
+                    status_callback,
+                    progress_callback,
+                    error_enabled_callback,
+                )
+
+        for slave_id_offset in threads:
+            exception = threads[slave_id_offset].exception()
```

Comment:
I don't understand how this loop works. `threads` is a dictionary, but what is `slave_id_offset` and why do we have to use it to get the value from `threads` again if we're already iterating over `threads`?

Reply:
This is to store the [Future](https://docs.python.org/3/library/concurrent.futures.html#future-objects) objects so I can check for any exceptions in any of the threads. I'll rename this dictionary so it is clearer.